[tests] メモリリーク確認スクリプト追加

nafuka11 opened the pull request:
Fix #70

## やったこと

### leaks.shの追加

メモリリーク確認用スクリプト `tests/leaks.sh` を追加しました。

#### 実行方法

```bash
cd tests
./leaks.sh
```

leaks.sh でやっていることは、 grademe.sh と似ています。
各ケースを minishell_leaks -C "コマンド" で実行し、leaksの結果をgrepし、OKかKOか判定しています。
確認することが違ったので grademe.sh とはファイルを分けています。leaksの結果を見るか、stdout, stderr, 終了ステータスの違いを見るかが違います。

#### テスト結果

OKな場合：緑色でokと出ます。
```bash
case: {ケース名} [ok]
```
KOな場合：赤色でkoと出ます。その後、leaksの結果が出力されます。
```bash
case: {ケース名} [ko]
Process {pid}: x leaks for x total leaked bytes.
    x (x bytes) ROOT LEAK: 0xXXXXXXX [x]
```
最後に `OKケース数/全体ケース数` が出力されます。

### leaks用実行ファイル minishell_leaks

`make leaks` でleaks用実行ファイル `minishell_leaks` を作成するようにしました。

`minishell_leaks` はプログラム終了時に system関数で leaks コマンドを呼びます。

これを実現するために、新たなdefine、 `LEAKS` を追加しました。
`LEAKS` が0以外の時は、GCCのattributeが追加されるようにしています。

既存の実行ファイル、`minishell` に影響がないことは `nm` コマンドで確認しています。

```bash
bash-3.2$ nm minishell | grep system
bash-3.2$ nm minishell_leaks | grep system
                 U _system
```
Cファイルのマクロについては、TRUE/FALSEから、1/0に変更しています。TRUE/FALSEだとt_boolのincludeが必要になったりするのでこうしました。

nafuka11 committed:
fix(const.h): DEBUGマクロをt_boolでなく0/1で持つよう変更

t_boolをincludeしてないとマクロが解釈されないため

nafuka11 committed:
feat(const.h): LEAKSマクロ追加

nafuka11 committed:
feat(main.c utils.h): LEAKSマクロ=1の時leaksコマンドを実行するように

nafuka11 committed:
feat(Makefile): make leaksでminishell_leaksを生成するように

nafuka11 committed:
chore(.gitignore): minishell_leaksを追加

nafuka11 committed:
test(leaks.sh): leaks確認用スクリプト追加

grademe.shと似た内容ではあるのですが別スクリプトとして追加しました

ToYeah reviewed (commented):
適当なfree部分を消して走らせてみるとちゃんとleakを検知してくれました！

nmの確認で問題ないことは理解しているのですが、main.cにsystem()が入っていることが少し引っかかります。
できれば別のファイルに分けたいなと思うのですが、動作に大きな影響が出たりしますか？

nafuka11 committed:
refactor: debug用マクロ（DEBUG, LEAKS）, 関数をdebug.hに移動

nafuka11 committed:
refactor: debug用関数end()をdebug.cに移動

nafuka11 commented:
別ファイルに分割しても動作に影響は出ないです！

`utils/debug.c` に分けることにしました。
また、debug用のヘッダファイル、`includes/debug.h` を作成し、デバッグ用のマクロを移しました。

現状、`utils/debug.c` は `make` 単体でもコンパイルされます。ただ `# if LEAKS` 内は展開されません。
コンパイルされるのがイマイチ……であれば、Makefileに `DEBUG_SRCS`, `DEBUG_OBJS` マクロを追加し、 `make srcs` に`DEBUG_SRCS` を除外するコマンドを追加して、除外することもできます。

ToYeah commented:
対応ありがとうございます！
できればsrcsにも入ってない状態がいいです...!

nafuka11 committed:
fix(Makefile): SRCSにdebug.cを含め内容修正

DEBUG_SRCS(debug.c), DEBUG_OBJSを追加
make leaksの時だけSRCSにDEBUG_SRCSを追加するように
make srcsでdebug.cを除外するように

nafuka11 commented:
SRCSにdebug.cを含めないよう修正しました！

ToYeah commented:
確認しました！
ヘッダのコンフリクトだけ直してもらえたらマージします！

nafuka11 committed:
Merge branch 'master' into 70-test-memory-leak

nafuka11 commented:
コンフリクト、修正しました！！
